pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,268 @@
+)
+
+// customTransformCollection digests CustomTransform objects and caches the results.
+type customTransformCollection struct {
+	// client is here for updating the status of a CustomTransform
+	client controlclient.CustomTransformInterface
+
+	// getTransformObjects is the part of the CustomTransform informer's cache.Indexer behavior
+	// that is needed here, for using the index named in `customTransformDomainIndexName`.
+	// It is used to get the CustomTransform objects relevant to a given GroupResource.
+	getTransformObjects func(indexName, indexedValue string) ([]any, error)
+
+	// enqueue is used to add a reference to a Binding that needs to be re-processed because
+	// of a change to a CustomTransform that the Binding is sensitive to.
+	enqueue func(any)
+
+	// mutex must be locked while accessing the following fields or their contents.
+	// The comments on the following fields and on groupResourceTransformData
+	// are things that hold true while the mutex is not locked.
+	mutex sync.Mutex
+
+	// grToTransformData has an entry for every GroupResource that some Binding cares about
+	// (i.e., lists an object of that GroupResource), and no more entries.
+	grToTransformData map[metav1.GroupResource]*groupResourceTransformData
+
+	// ctNameToSpec holds, for each CustomTranform whose spec contributed to an
+	// entry in grToTransformData, that CustomTransformSpec.
+	ctNameToSpec map[string]v1alpha1.CustomTransformSpec
+
+	// bindingNameToGroupResources tracks the set of GroupResource that each Binding
+	// references. This is so that when the set for a given Binding changes,
+	// for the GroupResources that are no longer in the set, the Binding's Name can
+	// be removed from groupResourceTransformData.bindingsThatCare.
+	// No Set[GroupResource] here is empty.
+	bindingNameToGroupResources map[string]sets.Set[metav1.GroupResource]
+}
+
+// groupResourceTransformData is the ingested custom transforms for a given GroupResource
+type groupResourceTransformData struct {
+	bindingsThatCare sets.Set[string /*Binding name*/] // not empty
+	ctNames          sets.Set[string /* CustomTransform name*/]
+	changes          customTransformChanges
+}
+
+type customTransformChanges struct {
+	removes []jsonpath.Query // immutable
+}
+
+func newCustomTransformCollection(client controlclient.CustomTransformInterface, getTransformObjects func(indexName, indexedValue string) ([]any, error), enqueue func(any)) *customTransformCollection {
+	return &customTransformCollection{
+		client:                      client,
+		getTransformObjects:         getTransformObjects,
+		enqueue:                     enqueue,
+		grToTransformData:           make(map[metav1.GroupResource]*groupResourceTransformData),
+		ctNameToSpec:                make(map[string]v1alpha1.CustomTransformSpec),
+		bindingNameToGroupResources: make(map[string]sets.Set[metav1.GroupResource]),
+	}
+}
+
+// GetCustomTransformData returns the groupResourceTransformData to use
```

Comment:
This documentation needs updating.

Reply:
Oh crud, I forgot to push the last change. Pushed just now.

Reply:
I pushed another commit, hopefully making the comments clearer.

The relation with Bindings is a critical part of the customTransformCollection functionality. It is only by tracking the set of relevant Bindings that the collection knows when to remove an entry for a particular GroupResource. I hope the documentation #2066 makes this clear: see the inclusion of the USES relation in the definition of the cache. At this point in time I would not recommend creating an abstraction whose only concrete realization is Binding; we can address that later if the need arises.